pixeltable/func/function.py

```diff
@@ -64,79 +73,189 @@ def display_name(self) -> str:
 
-    def call_return_type(self, kwargs: dict[str, Any]) -> ts.ColumnType:
+    def call_return_type(self, args: Sequence[Any], kwargs: dict[str, Any]) -> ts.ColumnType:
         """Return the type of the value returned by calling this function with the given arguments"""
+        assert not self.is_polymorphic
+        signature = self.signatures[0]
         if self._conditional_return_type is None:
-            return self.signature.return_type
-        bound_args = self.signature.py_signature.bind(**kwargs)
+            return signature.return_type
+        bound_args = signature.py_signature.bind(*args, **kwargs).arguments
         kw_args: dict[str, Any] = {}
         sig = inspect.signature(self._conditional_return_type)
         for param in sig.parameters.values():
-            if param.name in bound_args.arguments:
-                kw_args[param.name] = bound_args.arguments[param.name]
+            if param.name in bound_args:
+                kw_args[param.name] = bound_args[param.name]
         return self._conditional_return_type(**kw_args)
 
     def conditional_return_type(self, fn: Callable[..., ts.ColumnType]) -> Callable[..., ts.ColumnType]:
         """Instance decorator for specifying a conditional return type for this function"""
+        if len(self.signatures) > 1:
+            raise excs.Error('`conditional_return_type` is not supported for functions with multiple signatures')
         # verify that call_return_type only has parameters that are also present in the signature
         sig = inspect.signature(fn)
         for param in sig.parameters.values():
-            if param.name not in self.signature.parameters:
+            if param.name not in self.signatures[0].parameters:
                 raise ValueError(f'`conditional_return_type` has parameter `{param.name}` that is not in the signature')
         self._conditional_return_type = fn
         return fn
 
+    def overload(self, fn: Callable) -> Function:
+        raise NotImplementedError(f'Function of type {type(self)} does not support overloading')
+
     def using(self, **kwargs: Any) -> 'ExprTemplateFunction':
+        from .expr_template_function import ExprTemplateFunction, Template
+
+        assert len(self.signatures) > 0
+        if len(self.signatures) == 1:
+            # Only one signature: call _bind_to_template() and surface any errors directly
+            template = self._bind_to_template(kwargs)
+            return ExprTemplateFunction([template])
+        else:
+            # Multiple signatures: iterate over each signature and generate a template for each
+            # successful binding. If there are no successful bindings, raise a generic error.
+            # (Note that the resulting ExprTemplateFunction may have strictly fewer signatures than
+            # this Function, in the event that only some of the signatures are successfully bound.)
+            templates: list[Template] = []
+            for mono_fn in self.__resolutions:
```

Comment:
resolutions gets populated dynamically. How do we know we have seen the relevant resolutions here? Maybe the way that this is populated can be simplified, I'm finding it hard to follow (and why the current code is correct is not obvious to me).

Reply:
Just adding a note: with the most recent changes, once `using` has been called, subsequent `@overload` decorators will raise an error, so I think that answers this point.
<!-- Reviewable comment -OEWlntO2z5beybBETqL:b4jfjd2 -->
<!-- Sent from Reviewable.io -->